### ISSUE TITLE
Use dynamic variable as action action input values

### ISSUE THREAD

**Issue by then3rdman** [bitfocus/companion#1508]
I have seen a a few closes issues dismissing this request as it a per module feature however its not particularly useful when even the generic modules don't support it. 

Would it be possible to add some kind of interpolation to simply parse the variable to a string implicitly? As such the modules them self wouldn't need to care about the input as it would just be the value of the variable.

Failing that, can support for this be at least added to the generic modules?

**Comment by McHauge** [comment]
@then3rdman Hi, if it's the right way or not, then the first thing I would do is go to the generic modules and post a request for the specific support you want there, and then in the meantime, this can stay open as a general request.

From a module dev point of view, then I feel like there the knowledge and simplicity is missing when working with importing variables to inputs in actions and feedbacks, if this was a simple tig, that you could slap onto most input types that would be awesome. as an example, if you could just at the tag "allow-var: true" to text and number inputs fields that would be awesome.

**Comment by Julusian** [comment]
I would like to do this in the core of companion, but it needs quite a bit of thought.
Firstly, the ui for actions will need to be updated to be able to provide variables for the number and dropdown inputs without losing the number and dropdown ability.
A solution would be for each input field needs a checkbox to switch it to variables mode?

Even with this, I think every module will need to be aware that variables can be used. I know that some of my modules aren't guarding against bad data as well as they could because the input data is very controlled. eg, a number field being fed a variable which is currently `some text`. If modules are expecting a number it is highly likely that they will throw an error, which could crash companion. 

So at a minimum every module would need to be reviewed to ensure it has decent input data checking. Perhaps this is good motivation to put some work into trying to avoid errors crashing companion, but that is a hard task.

**Comment by then3rdman** [comment]
I think you could keep it much more simple if you just automatically interpolate the variables and pass them as the modules expected input. It would be nice for the UI to show feedback and what not but not immediately necessary. MVP would just be a simple conversation from variable name string, to value string

**Comment by Julusian** [comment]
Yeah I was meaning there that they would be converted before passing to the modules. The data types issues is still valid there

**Comment by McHauge** [comment]
@Julusian Yes the modules would need to add support/get updates in both cases, but having it as a simple addition makes it's a lot more sense to add, and go back and include. also, the option o NOT allow a variable would also make sense in some of the more complex and very specific cases. But I do agree you would need to have a check box of sort to switch from text/number to a searchable dropdown with variables.

and variables should be auto-converted to the input type, so a variable in a text field returns a string,  and in a number, field returns an integer. This would most likely also prompt a re-visite and a list of do's and dont's in variable naming and what values to use in them. as in to streamline them.

**Comment by Julusian** [comment]
@McHauge yeah, maybe they can be auto-converted to the input type. for numbers, they could be clamped to be within the valid range too, and if a dropdown, must be present in that list?
Another question is if the converted value is not valid, should we try and mangle to something that is valid, skip that action, or skip the whole press?

I am keen to try and get this done, but I have a lot of concerns about it :)

**Comment by SadowSon** [comment]
I think the simplest method would be to create a list of generic String variables within Companion.
For example... lets say you have 10 String Variables available by default. That string can be set to whatever you want.

BUTTON ONE:
On Button Press:
Send Input 1 to Preview
Save Input 1 name to String_Variable_1

BUTTON TWO
On Button Press:
Set Super Source Input 1 to String_Variable_1

Use Case:
The operator pushes Button One. This sends the input to the preview, that allows him to properly check that yes, this is the one he wants to send.
Then he pushes Button Two. That sends Input 1 to Super Source Input 1.

Technically, sending the input to the preview is entirely redundant - it's not actually required. It's only there as an action so you can confirm you're sending the right input. An operator programming the keyboard shouldn't need to worry about manually typing in the string command, because it's already setup to allow for saving the input name and then sending it.

You shouldn't need to mangle the name if it's not valid. The only thing that should be stored is the core input name. From there, when setting the input on the Super Source, the only thing it needs to know is which input it is, which should make for a fairly easy conversion.

You could probably also add a command that allows the user to set the string name manually themselves. Though attach a warning saying that allowing that isn't the intended function, and it's only so that commands can *technically* be passed between incompatible devices.

**Comment by McHauge** [comment]
just for reference this is my take on how it could be achieved: https://github.com/McHauge/companion-module-generic-variables

just compile a build locally with that module and you will be able to add them as a module :)

**Comment by Julusian** [comment]
@McHauge only helps if the module you wish to feed the variable supports it :) I havent forgotten, that idea is still on my list of things to try and write as part of companion

> @SadowSon which should make for a fairly easy conversion.

Maybe, depending on what name you put in the variable. Would you put `Camera1` `Input1` `In1` `Cam1` `1` `Camera 1` or the name you have assigned in that variable? 
On my atem right now input 1 is called `Camera 12`, so if `Camera 12` is seen should that be interpreted as 1 or 12? If 1, why? All the variables the atem module currently provides will use `Camera 12` in their naming because that is what you would want written on buttons.
Or for bars, would you want to write `Color Bars` `Colour Bars` `Bars` `1000`. There are so many different ways each input could be written, some of which become rather ambiguous.

The answer here is probably to only accept the numeric values, and duplicate a load of the variables in the atem module to have numeric versions. But that will mean that you will need to be aware of the input ids of the inputs when setting variables, which isnt very friendly


> You could probably also add a command that allows the user to set the string name manually themselves. Though attach a warning saying that allowing that isn't the intended function, and it's only so that commands can technically be passed between incompatible devices.

I dont follow what you mean here.

**Comment by SadowSon** [comment]
Input 1 should be saved in the string variable as Input1 (Or whatever the internal name for it is in the module)
The display name that you've assigned to that, in this case "Camera 12", is largely irrelevant. A name can be anything, and changes between every different event you do. On the flip side, system input names never change. Input 1 will always be Input 1.  Input 2 is always Input 2. The display name attached to it doesn't matter, as the variable data doesn't even need to be read by the operator.
So long as they say to save the name of Input 1 (In this case, being Input1 and not Camera 12) into Variable 1, then when you tell  your output or Super Source to set it's input, It can interpret that Input1 is a valid input when it sends the command to the switcher.
That cuts out the entirety of the ambiguity of the naming.

**Comment by 4ndrw** [comment]
I'd totally use this....

Currently creating lots of near identical buttons with slightly different sysex strings for different commands or the same commands but on other midi channels for a multi timbral synthesiser...

Some variables inside the textbox for the command I'm sending, eg. for currently selected midi channel, parameter ID or parameter value... would be awesome!

That said, I'm excited at what I can do already in this software

**Comment by Julusian** [comment]
#2345 proposes a more detailed plan for this, closing as a duplicate of that
